sirepo/pkcli/runner.py

```diff
@@ -0,0 +1,222 @@
+        # bound, perhaps by clarifying the split in responsibilities between
+        # the on-disk simulation_db versus the in-memory status.
+        self.jobs = {}
+        self._nursery = nursery
+
+    async def start_job(self, jid, config):
+        await self._nursery.start(self._run_job, jid, config)
+
+    async def _run_job(
+            self, jid, run_dir, config, *, task_status=trio.TASK_STATUS_IGNORED
+    ):
+        # XX TODO: there are all kinds of race conditions here, e.g. if the
+        # same jid gets started multiple times in parallel... we need to
+        # revisit this once jids are globally unique, and possibly add
+        # features like "if there is another job running with the same
+        # user/sim/job but a different hash, then auto-cancel that other job"
+        self.jobs[jid] = _JobInfo(jid, run_dir, _JobStatus.RUNNING)
+        try:
+            container = await trio.run_sync_in_worker_thread(
+                functools.partial(
+                    _DOCKER.containers.run,
+                    config['image'],
+                    config['command'],
+                    working_dir=config['working_dir'],
+                    # {host path: {'bind': container path, 'mode': 'rw'}}
+                    volumes=config['volumes'],
+                    name=jid,
+                    auto_remove=True,
+                    detach=True,
+                    init=True,
+                )
+            )
+            task_status.started()
+            # Returns a dict like: {'Error': None, 'StatusCode': 0}
+            # XX TODO but there may be a race condition where if the container
+            # finishes before our call to wait() starts, then it just errors
+            # out with docker.errors.NotFound
+            with _catch_and_log_errors(
+                    Exception, 'error waiting for container {}', jid
+            ):
+                pkdc(await _container_wait(container))
+        finally:
+            self.jobs[jid] = _JobStatus.FINISHED
+
+
+async def _start_job(job_tracker, request):
+    pkdc('start_job: {}', request)
+    await job_tracker.start_job(
+        request['jid'], request['run_dir'], request['config']
+    )
+    return {}
+
+
+async def _job_status(job_tracker, request):
+    pkdc('job_status: {}', request)
+    job_info = job_tracker.jobs.get(request['jid'])
+    if job_info is None:
+        return {"status": _JobStatus.NOT_STARTED.value}
+    else:
+        return {"status": job_info.status.value}
+
+
+async def _cancel_job(job_tracker, request):
+    job_info = job_tracker.jobs[request['jid']]
+    if job_info.status is not JobStatus.RUNNING:
+        return {}
+    simulation_db.write_result({'state': 'canceled'}, run_dir=job_info.run_dir)
+    # XX TODO: this is what api_runCancel used to do, but we can't really do
+    # it here. What should we do? (For a cancelled job, should we just delete
+    # the run-dir entirely, once we have the full hash in the jid?)
```

Comment:
canceled jobs can contain usable data. When we run parallel simulations, the user can cancel at any time, and still have the results already computed. Perhaps this might be a different state.